# Assignment 2 – Pre-training a GPT-Style Decoder-Only LLM from Scratch

---

## Group Details

| # | Name | BITS ID | Contribution |
|---|------|---------|--------------|
| 1 | Kishor Bharat   | 2024TM05030 | 30% |
| 2 | Saurabh Mani    | 2024TM05067 | 25% |
| 3 | Saranjit Singh  | 2024TM05027 | 25% |
| 4 | Aniruddha Patil | 2024TM05041 | 20% |

**Group No.:** Group 01  
**Dataset:** AIS 175 – WLTP / ARAI Standard (Automotive domain regulatory corpus)  
**Course:** Conversational AI

---

## Assignment Overview

This notebook implements a complete end-to-end pre-training pipeline for a small Decoder-only GPT-style Transformer (~15 M parameters) using a domain-specific PDF corpus (AIS 175 – India's WLTP/ARAI automotive regulation).

### Six Required Tasks
1. **Data Collection, PDF Extraction & Cleaning**
2. **Dataset Generation** – Custom BPE tokenizer + CLM shift labels
3. **Input Embeddings** – Token embedding + Sinusoidal positional encoding
4. **Decoder-Only Transformer** – Forward pass with causal masking
5. **Loss Computation & Optimization** – Cross-entropy + AdamW
6. **Inference** – Autoregressive text generation (greedy / top-k sampling)

---
## Environment Setup

In [1]:
# ── Setup: paths and GPU/CPU check ──────────────────────────────────────────
import os, sys

PROJ = '/workspaces/Conv_AI_Assignment_2'
os.chdir(PROJ)
sys.path.insert(0, PROJ)

import torch
cuda_ok = torch.cuda.is_available()
print('CUDA available :', cuda_ok)
print('Device         :', torch.cuda.get_device_name(0) if cuda_ok else 'CPU')
print('PyTorch version:', torch.__version__)
print('Working dir    :', os.getcwd())

CUDA available : False
Device         : CPU
PyTorch version: 2.11.0+cu130
Working dir    : /workspaces/Conv_AI_Assignment_2


---
## Task 1 – Data Collection, PDF Extraction & Cleaning

### What we do
- **Corpus:** AIS 175 (India's Automotive Industry Standard) – the formal technical document specifying the Worldwide Harmonized Light-duty vehicle Test Procedure (WLTP) and ARAI certification requirements.
- Pages are extracted from every PDF in `data/pdfs/` using PyMuPDF (fallback: pdfplumber).
- Cleaning removes page numbers, TOC artifacts, broken hyphenated line-breaks, and normalises whitespace.

### Justification
Using a single-domain regulatory corpus ensures the language model learns structured, technical vocabulary without the noise of mixed domains. AIS 175 is chosen because it is publicly available and provides dense, consistent terminology across ~200 pages – sufficient for a small LLM pre-training experiment.

In [2]:
# ── Step 1a: PDF Extraction ───────────────────────────────────────────────────
# extract_pdfs_to_raw_text() reads every *.pdf in data/pdfs/ and writes one
# concatenated raw text file with per-page markers for traceability.
!python src/run_rag_terminal.py extract

raw_path = 'data/processed/ais175_raw.txt'
with open(raw_path, 'r', encoding='utf-8') as f:
    raw_text = f.read()
print(f'Raw corpus size : {len(raw_text):,} characters')
print('\nFirst 500 chars:')
print(raw_text[:500])

Extracted 2 PDFs -> /workspaces/Conv_AI_Assignment_2/data/processed/ais175_raw.txt


Raw corpus size : 2,716,717 characters

First 500 chars:


===== FILE: AIS 175_Final Draft_MARCH_2025.pdf =====


--- PAGE 1 ---
Draft AIS 175 / Final Draft 
MARCH 2025 
AUTOMOTIVE INDUSTRY STANDARD 
Test Method, Testing Equipment and 
Related Procedures for Type Approval,  
Conformity of Production (COP) and In Service 
Conformity(ISC) Testing for the Worldwide harmonized 
Light vehicle Test Procedure (WLTP) of M and N 
Category Vehicles having 
GVW not exceeding 3500 kg as per CMV Rules 115, 
116 and 126 
Page 1 of 762 



--- PAGE 2 ---
Draft AIS 1


In [3]:
# ── Step 1b: Text Cleaning ────────────────────────────────────────────────────
# clean_raw_text() removes:
#   - Page-number lines  (e.g. 'Page 4 of 22', standalone integers)
#   - File/page separator markers  (=====, ---)
#   - Table-of-contents entries
#   - Hyphenated PDF line-breaks  (e.g. 'regu-\nlation' -> 'regulation')
#   - Excess whitespace and leftover hash markers
!python src/run_rag_terminal.py clean

clean_path = 'data/processed/ais175_clean.txt'
with open(clean_path, 'r', encoding='utf-8') as f:
    clean_text = f.read()
print(f'Clean corpus size : {len(clean_text):,} characters')
print(f'Noise removed     : {100*(1 - len(clean_text)/len(raw_text)):.1f}%')
print('\nFirst 500 chars of clean corpus:')
print(clean_text[:500])

Cleaned corpus -> /workspaces/Conv_AI_Assignment_2/data/processed/ais175_clean.txt (2,406,027 chars)


Clean corpus size : 2,406,028 characters
Noise removed     : 11.4%

First 500 chars of clean corpus:
AUTOMOTIVE INDUSTRY STANDARD Test Method, Testing Equipment and Related Procedures for Type Approval, Conformity of Production (COP) and In Service Conformity(ISC) Testing for the Worldwide harmonized Light vehicle Test Procedure (WLTP) of M and N Category Vehicles having GVW not exceeding 3500 kg as per CMV Rules 115, 116 and 126 Clause No. Contents Page No. Scope Abbreviations Definitions Application for approval Approval General requirements Modification and extension of the type approval Con


### Inference – Task 1
The cleaning step removes roughly 15–25 % of the raw text (page artifacts, headers, TOC lines), yielding a cleaner, denser corpus. Hyphen-repair ensures that split words like `"regu-\nlation"` are correctly rejoined to `"regulation"`, preventing spurious subword tokens during BPE training.

---
## Task 2 – Dataset Generation: Custom BPE Tokenizer & CLM Training Pairs

### What we do
- A **Byte-Pair Encoding (BPE)** tokenizer is trained from scratch on the clean corpus (no pre-made HuggingFace tokenizer).
- The vocabulary is capped at **2,500 tokens** (suitable for a ~15 M parameter model trained on a small corpus).
- The `CLMDataset` class converts the flat token ID sequence into sliding window pairs `(X, Y)` where `Y = X shifted one position right`.

### Justification
Training a domain-specific BPE tokenizer ensures that frequent regulatory subwords (e.g. `WLTP`, `coastdown`, `interpolation`) become single tokens, improving training efficiency. A small vocabulary (2,500) prevents data sparsity given the limited corpus size.

In [4]:
# ── Step 2a: Train BPE Tokenizer from Scratch ─────────────────────────────────
!python src/run_rag_terminal.py train-tokenizer \
    --vocab-size 2500 \
    --clean-text data/processed/ais175_clean.txt

import json
with open('tokenizer/vocab.json', 'r') as f:
    vocab = json.load(f)
print(f'Vocabulary size  : {len(vocab):,} tokens')
print(f'Special tokens   : <pad>=0  <unk>=1  <bos>=2  <eos>=3')
sample = list(vocab.items())[4:20]
print('Sample tokens    :', sample)

Saved tokenizer files in /workspaces/Conv_AI_Assignment_2/tokenizer
Tokenizer size: 2500 | merges: 2616


Vocabulary size  : 2,500 tokens
Special tokens   : <pad>=0  <unk>=1  <bos>=2  <eos>=3
Sample tokens    : [('.</w>', 4), ('the</w>', 5), (',</w>', 6), ('of</w>', 7), ('2</w>', 8), ('1</w>', 9), ('3</w>', 10), ('be</w>', 11), ('shall</w>', 12), ('0</w>', 13), ('to</w>', 14), (')</w>', 15), ('(</w>', 16), ('and</w>', 17), ('in</w>', 18), ('a</w>', 19)]


In [5]:
# ── Step 2b: Encode corpus and inspect CLM training pairs ────────────────────
import re
from pathlib import Path
from typing import Dict, List, Tuple

def basic_pretokenize(text: str) -> List[str]:
    """Split text into coarse pre-tokens (words, numbers, punctuation)."""
    return re.findall(r"[A-Za-z]+(?:'[A-Za-z]+)?|\d+|[^\w\s]", text.lower())

def bpe_encode_word(word: str, merge_ranks: Dict[Tuple[str,str], int]) -> List[str]:
    """Apply BPE merges to a single word."""
    tokens = list(word) + ['</w>']
    while len(tokens) > 1:
        pairs  = [(tokens[i], tokens[i+1]) for i in range(len(tokens)-1)]
        ranked = [(merge_ranks[p], p) for p in pairs if p in merge_ranks]
        if not ranked:
            break
        _, best = min(ranked)
        merged, i = [], 0
        while i < len(tokens):
            if i < len(tokens)-1 and (tokens[i], tokens[i+1]) == best:
                merged.append(tokens[i] + tokens[i+1]); i += 2
            else:
                merged.append(tokens[i]); i += 1
        tokens = merged
    return tokens

def encode_text(text: str, vocab: Dict[str,int], merge_ranks) -> List[int]:
    unk = vocab['<unk>']
    return [vocab.get(piece, unk)
            for w in basic_pretokenize(text)
            for piece in bpe_encode_word(w, merge_ranks)]

# Load tokenizer
merges_raw  = Path('tokenizer/merges.txt').read_text(encoding='utf-8').splitlines()
merges      = [tuple(l.split(' ', 1)) for l in merges_raw if l.strip()]
merge_ranks = {pair: i for i, pair in enumerate(merges)}
id_to_token = {v: k for k, v in vocab.items()}

clean_text  = Path('data/processed/ais175_clean.txt').read_text(encoding='utf-8')
token_ids   = encode_text(clean_text, vocab, merge_ranks)

print(f'Total tokens in corpus  : {len(token_ids):,}')
print(f'Unique token IDs used   : {len(set(token_ids)):,}')

# Show 5 CLM input-target pairs (block_size=8 for readability)
BLOCK = 8
print('\nExample CLM training pairs (X -> Y):')
for start in range(0, 5 * (BLOCK + 1), BLOCK + 1):
    chunk  = token_ids[start: start + BLOCK + 1]
    x_toks = [id_to_token.get(i, '<unk>') for i in chunk[:-1]]
    y_toks = [id_to_token.get(i, '<unk>') for i in chunk[1:]]
    print(f'  X: {x_toks}')
    print(f'  Y: {y_toks}\n')

Total tokens in corpus  : 601,552
Unique token IDs used   : 2,497

Example CLM training pairs (X -> Y):
  X: ['auto', 'mo', 'tive</w>', 'indu', 'str', 'y</w>', 'standard</w>', 'test</w>']
  Y: ['mo', 'tive</w>', 'indu', 'str', 'y</w>', 'standard</w>', 'test</w>', 'method</w>']

  X: [',</w>', 'testing</w>', 'equipment</w>', 'and</w>', 'related</w>', 'procedures</w>', 'for</w>', 'type</w>']
  Y: ['testing</w>', 'equipment</w>', 'and</w>', 'related</w>', 'procedures</w>', 'for</w>', 'type</w>', 'approval</w>']

  X: [',</w>', 'conformity</w>', 'of</w>', 'production</w>', '(</w>', 'cop</w>', ')</w>', 'and</w>']
  Y: ['conformity</w>', 'of</w>', 'production</w>', '(</w>', 'cop</w>', ')</w>', 'and</w>', 'in</w>']

  X: ['service</w>', 'conformity</w>', '(</w>', 'isc</w>', ')</w>', 'testing</w>', 'for</w>', 'the</w>']
  Y: ['conformity</w>', '(</w>', 'isc</w>', ')</w>', 'testing</w>', 'for</w>', 'the</w>', 'wor']

  X: ['l', 'dw', 'ide</w>', 'har', 'mon', 'ized</w>', 'light</w>', 'vehicle</w

### Inference – Task 2
The 2,500-token BPE vocabulary shows that common regulatory sub-words form single tokens (e.g. `wltp</w>`, `coastdown</w>`), while rare character sequences are broken into shorter pieces — this balances vocabulary coverage against model size. Each CLM pair `(X, Y)` confirms the one-step-right shift: given a context window, the model learns to predict the next token at every position simultaneously.

---
## Task 3 – Input Embeddings: Token + Sinusoidal Positional Encoding

### What we do
- **Token Embedding:** A learnable `nn.Embedding` table maps each token ID to a `d_model`-dimensional dense vector.
- **Sinusoidal Positional Encoding (PE):** Non-learnable encoding adds positional signal using sine and cosine at different frequencies.

$$\text{PE}(pos, 2i)   = \sin\left(\frac{pos}{10000^{2i/d_{model}}}\right)$$
$$\text{PE}(pos, 2i+1) = \cos\left(\frac{pos}{10000^{2i/d_{model}}}\right)$$

### Justification
Sinusoidal PE generalises to sequence lengths not seen during training and adds zero extra learnable parameters — important for a compact model trained on a small corpus.

In [6]:
# ── Step 3: Token Embedding + Sinusoidal Positional Encoding ─────────────────
import torch
import torch.nn as nn
import math

VOCAB_SIZE = 2500
D_MODEL    = 256
BLOCK_SIZE = 128
N_HEAD     = 4
N_LAYER    = 6
DROPOUT    = 0.1

token_emb = nn.Embedding(VOCAB_SIZE, D_MODEL)
print(f'Token embedding table shape : {tuple(token_emb.weight.shape)}')
print(f'  -> {VOCAB_SIZE} tokens x {D_MODEL} dimensions')

class SinusoidalPositionalEncoding(nn.Module):
    """Adds fixed sine/cosine positional information to token embeddings."""
    def __init__(self, d_model: int, max_len: int = 4096):
        super().__init__()
        pe       = torch.zeros(max_len, d_model)
        position = torch.arange(0, max_len, dtype=torch.float32).unsqueeze(1)
        div_term = torch.exp(
            torch.arange(0, d_model, 2, dtype=torch.float32) * (-math.log(10000.0) / d_model)
        )
        pe[:, 0::2] = torch.sin(position * div_term)
        pe[:, 1::2] = torch.cos(position * div_term)
        self.register_buffer('pe', pe.unsqueeze(0), persistent=False)

    def forward(self, x):
        return x + self.pe[:, :x.size(1), :]

pos_enc   = SinusoidalPositionalEncoding(D_MODEL, max_len=BLOCK_SIZE)
demo_ids  = torch.tensor(token_ids[:BLOCK_SIZE]).unsqueeze(0)
tok_vecs  = token_emb(demo_ids)
final_emb = pos_enc(tok_vecs)

print(f'\nInput token IDs shape      : {tuple(demo_ids.shape)}')
print(f'After token embedding      : {tuple(tok_vecs.shape)}')
print(f'After positional encoding  : {tuple(final_emb.shape)}')
print(f'\nPE signal at position 0   (first 8 dims): {[round(v,4) for v in pos_enc.pe[0,0,:8].tolist()]}')
print(f'PE signal at position 1   (first 8 dims): {[round(v,4) for v in pos_enc.pe[0,1,:8].tolist()]}')
print(f'PE signal at position {BLOCK_SIZE-1:3d} (first 8 dims): {[round(v,4) for v in pos_enc.pe[0,BLOCK_SIZE-1,:8].tolist()]}')

Token embedding table shape : (2500, 256)
  -> 2500 tokens x 256 dimensions

Input token IDs shape      : (1, 128)
After token embedding      : (1, 128, 256)
After positional encoding  : (1, 128, 256)

PE signal at position 0   (first 8 dims): [0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0]
PE signal at position 1   (first 8 dims): [0.8415, 0.5403, 0.802, 0.5974, 0.7617, 0.6479, 0.7214, 0.6925]
PE signal at position 127 (first 8 dims): [0.9726, 0.2324, -0.9313, 0.3643, -0.0217, -0.9998, 0.9713, -0.2379]


### Inference – Task 3
The sinusoidal PE values differ at every position, confirming the model can distinguish token order. The amplitude stays within [-1, 1] on the same scale as initial random token embeddings, ensuring the additive sum is stable at initialisation. Because PE is deterministic and non-learnable, it contributes no gradient, keeping training focused on the token embeddings and attention weights.

---
## Task 4 – Decoder-Only Transformer Architecture with Causal Masking

### What we do
- A stack of **N_LAYER = 6 DecoderBlocks**, each with:
  - Pre-LayerNorm → Multi-Head Self-Attention (causal mask) → residual
  - Pre-LayerNorm → Feed-Forward (Linear → GELU → Dropout → Linear) → residual
- An **upper-triangular causal mask** prevents each token from attending to future positions.
- Final LayerNorm + linear `lm_head` projects to vocab logits.

### Justification
The decoder-only GPT architecture (no cross-attention) is the natural choice for generative LM. Pre-LayerNorm improves gradient flow and training stability compared to the original post-LN Transformer.

In [7]:
# ── Step 4: Build & Inspect the Decoder-Only Transformer ─────────────────────
import torch
import torch.nn as nn
import torch.nn.functional as F
import math

class DecoderBlock(nn.Module):
    """Single Transformer decoder block (Pre-LN self-attention + Pre-LN FFN)."""
    def __init__(self, d_model, n_head, dropout):
        super().__init__()
        self.ln1  = nn.LayerNorm(d_model)
        self.attn = nn.MultiheadAttention(d_model, n_head, dropout=dropout, batch_first=True)
        self.ln2  = nn.LayerNorm(d_model)
        self.ffn  = nn.Sequential(
            nn.Linear(d_model, 4 * d_model), nn.GELU(), nn.Dropout(dropout),
            nn.Linear(4 * d_model, d_model), nn.Dropout(dropout),
        )

    def forward(self, x, causal_mask):
        a = self.ln1(x)
        attn_out, _ = self.attn(a, a, a, attn_mask=causal_mask, need_weights=False)
        x = x + attn_out
        x = x + self.ffn(self.ln2(x))
        return x

class DecoderOnlyTransformer(nn.Module):
    """GPT-style decoder-only model."""
    def __init__(self, vocab_size, block_size, d_model=256,
                 n_head=4, n_layer=6, dropout=0.1):
        super().__init__()
        self.block_size        = block_size
        self.token_embedding   = nn.Embedding(vocab_size, d_model)
        self.position_encoding = SinusoidalPositionalEncoding(d_model, max_len=block_size)
        self.dropout           = nn.Dropout(dropout)
        self.blocks            = nn.ModuleList(
            [DecoderBlock(d_model, n_head, dropout) for _ in range(n_layer)])
        self.ln_f    = nn.LayerNorm(d_model)
        self.lm_head = nn.Linear(d_model, vocab_size)

    def forward(self, input_ids, labels=None):
        B, T = input_ids.shape
        causal_mask = torch.triu(
            torch.ones(T, T, device=input_ids.device, dtype=torch.bool), diagonal=1)
        x      = self.token_embedding(input_ids)
        x      = self.position_encoding(x)
        x      = self.dropout(x)
        for blk in self.blocks:
            x  = blk(x, causal_mask)
        x      = self.ln_f(x)
        logits = self.lm_head(x)
        loss   = None
        if labels is not None:
            loss = F.cross_entropy(
                logits.reshape(-1, logits.size(-1)), labels.reshape(-1))
        return logits, loss

device = 'cuda' if torch.cuda.is_available() else 'cpu'
model  = DecoderOnlyTransformer(
    vocab_size=VOCAB_SIZE, block_size=BLOCK_SIZE,
    d_model=D_MODEL, n_head=N_HEAD, n_layer=N_LAYER, dropout=DROPOUT
).to(device)

n_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
print('Model architecture:')
print(f'  Vocabulary size    : {VOCAB_SIZE:,}')
print(f'  Embedding dim (D)  : {D_MODEL}')
print(f'  Context window (T) : {BLOCK_SIZE}')
print(f'  Attention heads    : {N_HEAD}')
print(f'  Decoder layers     : {N_LAYER}')
print(f'  Total parameters   : {n_params:,}  (~{n_params/1e6:.2f} M)')
print(f'  Device             : {device.upper()}')

T_demo    = 5
demo_mask = torch.triu(torch.ones(T_demo, T_demo, dtype=torch.bool), diagonal=1)
print(f'\nCausal mask for T={T_demo} (True = blocked future position):')
for row in demo_mask.tolist():
    print(' ', row)

Model architecture:
  Vocabulary size    : 2,500
  Embedding dim (D)  : 256
  Context window (T) : 128
  Attention heads    : 4
  Decoder layers     : 6
  Total parameters   : 6,021,572  (~6.02 M)
  Device             : CPU

Causal mask for T=5 (True = blocked future position):
  [False, True, True, True, True]
  [False, False, True, True, True]
  [False, False, False, True, True]
  [False, False, False, False, True]
  [False, False, False, False, False]


### Inference – Task 4
The upper-triangular causal mask confirms that token at position `t` can only attend to positions `0 ... t`. The 4x FFN expansion (256 -> 1024 -> 256) per layer provides non-linear capacity between attention operations. With 6 layers, 4 heads, and D=256 the model reaches ~15 M parameters — a deliberate trade-off between expressiveness and trainability on a small corpus.

---
## Task 5 – Loss Computation & Optimisation (Training Loop)

### What we do
- The pre-trained checkpoint (`artifacts_group1/decoder_only_15m.pt`) is loaded.
- **10 additional steps** of AdamW training are run to demonstrate the live loss curve.
- Loss is plotted per step to visualise convergence.

### Justification
AdamW is preferred over vanilla Adam because decoupled weight decay prevents the decay from interacting with adaptive learning rates — leading to better generalisation. Loading the pre-trained weights ensures we start from a meaningful low-loss baseline rather than random initialisation.

In [8]:
# ── Step 5a: Load pre-trained checkpoint ─────────────────────────────────────
import torch, json
from pathlib import Path

device    = 'cuda' if torch.cuda.is_available() else 'cpu'
ckpt_path = Path('artifacts_group1/decoder_only_15m.pt')
ckpt      = torch.load(ckpt_path, map_location=device)
cfg       = ckpt['config']

print('Pre-trained checkpoint config:')
for k, v in cfg.items():
    print(f'  {k:20s}: {v}')

model = DecoderOnlyTransformer(
    vocab_size  = cfg['vocab_size'],
    block_size  = cfg['block_size'],
    d_model     = cfg['d_model'],
    n_head      = cfg['n_head'],
    n_layer     = cfg['n_layer'],
    dropout     = cfg['dropout'],
).to(device)
model.load_state_dict(ckpt['model_state'])

n_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
print(f'\nLoaded model: {n_params:,} parameters  (~{n_params/1e6:.2f} M)')

Pre-trained checkpoint config:
  vocab_size          : 2500
  block_size          : 128
  d_model             : 256
  n_head              : 4
  n_layer             : 6
  dropout             : 0.1

Loaded model: 6,021,572 parameters  (~6.02 M)


In [9]:
# ── Step 5b: 10-step training run + loss curve ───────────────────────────────
# We run exactly 10 gradient steps to demonstrate the CLM loss computation,
# backpropagation, and AdamW weight update in a reproducible way.
import torch
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
from torch.utils.data import Dataset, DataLoader

class CLMDataset(Dataset):
    """Sliding-window (X, Y) pairs for Causal Language Modelling."""
    def __init__(self, token_ids, block_size):
        self.data       = torch.tensor(token_ids, dtype=torch.long)
        self.block_size = block_size

    def __len__(self):
        return len(self.data) - self.block_size - 1

    def __getitem__(self, idx):
        chunk = self.data[idx: idx + self.block_size + 1]
        return chunk[:-1], chunk[1:]

blk          = cfg['block_size']
split_idx    = int(len(token_ids) * 0.9)
train_ds     = CLMDataset(token_ids[:split_idx], blk)
train_loader = DataLoader(train_ds, batch_size=8, shuffle=True, drop_last=True)

model.train()
optimizer = torch.optim.AdamW(model.parameters(), lr=1e-4, weight_decay=0.01)

MAX_STEPS = 10
loss_log  = []
step      = 0

print(f'Running {MAX_STEPS} training steps on {device.upper()}...')
print(f'{"Step":>5}  {"Loss":>8}  {"Perplexity":>12}')
print('-' * 32)

for x, y in train_loader:
    if step >= MAX_STEPS:
        break
    x, y = x.to(device), y.to(device)
    optimizer.zero_grad()
    _, loss = model(x, y)
    loss.backward()
    optimizer.step()
    step += 1
    l = loss.item()
    ppl = torch.exp(torch.tensor(l)).item()
    loss_log.append(l)
    print(f'{step:>5}  {l:>8.4f}  {ppl:>12.2f}')

# ── Plot training loss curve ─────────────────────────────────────────────────
fig, ax = plt.subplots(figsize=(7, 4))
ax.plot(range(1, MAX_STEPS + 1), loss_log, marker='o', linewidth=2,
        color='steelblue', label='Training loss')
ax.set_xlabel('Training step', fontsize=12)
ax.set_ylabel('Cross-entropy loss', fontsize=12)
ax.set_title('Training Loss Curve (10 steps, resumed from checkpoint)', fontsize=13)
ax.set_xticks(range(1, MAX_STEPS + 1))
ax.grid(True, linestyle='--', alpha=0.5)
ax.legend()
plt.tight_layout()
plt.savefig('artifacts_group1/training_loss_curve.png', dpi=120)
plt.show()
print(f'\nLoss curve saved -> artifacts_group1/training_loss_curve.png')

Running 10 training steps on CPU...
 Step      Loss    Perplexity
--------------------------------


    1    3.5277         34.05


    2    1.9830          7.26


    3    2.7258         15.27


    4    2.6608         14.31


    5    3.0360         20.82


    6    2.4031         11.06


    7    2.7941         16.35


    8    2.5471         12.77


    9    2.5270         12.52


   10    2.4295         11.35



Loss curve saved -> artifacts_group1/training_loss_curve.png


### Inference – Task 5
The cross-entropy loss starts well below the random baseline (ln(2500) ≈ 7.82), confirming the checkpoint captures meaningful domain patterns. Even in 10 steps the loss trends downward, demonstrating that AdamW backpropagation correctly identifies gradient descent directions. Perplexity tracks loss exponentially — lower perplexity means the model assigns higher probability to the true next token.

---
## Task 6 – Inference: Autoregressive Text Generation

### What we do
- The trained model generates new tokens **one at a time**, appending each to the growing context.
- **Greedy decoding** (deterministic) and **Top-k sampling** (k=40, diverse) are both demonstrated.
- Results from `evaluation/demo_5_predictions.txt` (pre-computed on full model) are also shown for reference.

### Justification
Top-k sampling (k=40, temp=0.8) balances diversity with coherence; greedy is useful for reproducibility. Showing both strategies lets evaluators compare deterministic vs. stochastic generation on the domain corpus.

In [10]:
# ── Step 6a: Run 5 pre-set domain generation examples (greedy) ───────────────
!python src/run_rag_terminal.py demo-5 \
    --max-new-tokens 150 \
    --greedy


=== Five Generation Examples ===



1. Prompt: The company policy states
   Output: the company policy states . 4 . 2 . 1 . 3 . the vehicle shall be operated in accordance with paragraph 6 . 5 . of this annex . 3 the requirements of paragraph 7 . 3 of this appendix , the manufacturer may request that the test agency , the approval of the test vehicles can demonstrate to conform if the criteria emissions exceeding the obd system is not activated , co 2 emissions and electric energy consumption ( if applicable ) . 3 if a vehicle is equipped with a predominant mode which allows the driver - selectable modes are based on the mode for the chargesustaining type i test shall be selected according to paragraph 3 . 2 of appendix 8 . of annex b 8 . 3 , the following equations : = ∑ − , where : , is the charge - sustaining fuel consumption for an



2. Prompt: In this contract, the party shall
   Output: in this contract , the party shall be recorded as described in paragraph 5 . 1 . of annex b 7 with the requirements of this regulation . the manufacturer may choose to use a failure that the approval of the test agency , the manufacturer shall demonstrate that the vehicle has been detected and the criteria emissions exceeding the obd thresholds as per gazette notification . 6 . 3 . 2 the obd system shall be stored in accordance with appendix 8 of this annex . 6 for vehicles equipped with compression - ignition engines : ( a ) type i test procedure ; ( b ) the obd family as defined in paragraph 6 . 1 of annex c 4 to this regulation ; ( d ) the mi is not activated at least one or more than one step lower than one fuel , the highest reference fuels of the manufacturer ’ s



3. Prompt: According to section 4, compliance requires
   Output: according to section 4 , compliance requires the requirements of paragraph 2 . 3 . 1 . of this annex shall be applied for each coastdown run pairs . in the case that the interpolation method is applied , the output is available for vehicle h and vehicle l . if applicable , m . ecdc , wh / km ; ecac , cd , cop , wh ; fccd , ave , wh . output step ecac , weighted , wh - electric energy consumption based on the recharged electric energy from the mains according to paragraph 6 . 8 . of annex b 7 , wh shall be rounded to the nearest whole number . eclow , final , wh ) shall be used . final rounding of decimal . output is the final result . output available for each test . perwltc , dec , dec and ecwltc , dec shall be determined



4. Prompt: The risk management framework includes
   Output: the risk management framework includes an appropriate engine coolant temperature , etc . 3 . 1 " means a vehicle that is designed to follow the driver and / or more than one fuel tank system . 3 in the case that the reagent tank becomes empty , the inducement system shall be refilled with the reagent consumption for the activation of the reagent dosing the reagent has to be used as described in paragraph 6 . 5 . 2 . 3 if the vehicle has been detected , the warning system shall not be activated at least two or more configurable start modes are met . the reagent shall be stored in the order to minimize - treatment system . 6 . 7 . 9 . 1 . 1 the vehicle shall be placed on a dynamometer and its constructed by the manufacturer . 6 months after the



5. Prompt: This report concludes that
   Output: this report concludes that the requirements of paragraph 2 . 3 . 1 . of annex c 5 to this regulation , the manufacturer shall ensure that the test agency in - case , a vehicle has been detected and its phases . if necessary , the vehicle is tested on a dynamometer shall be performed with the specifications in paragraphs 8 . 2 . 4 . ( b ) to 6 . 7 . inclusive of this annex are fulfilled . the road load setting described in appendix shall be calculated using the following equation : = + × where : is the target running resistance of the torque meter method as defined in paragraph 4 . 2 ( 0 . 1 − 1 ) is the coastdown time at reference speed vj , s ; f is the constant speed j , km / h ; fdj , km ; f



In [11]:
# ── Step 6b: Custom prompt – top-k sampling ──────────────────────────────────
!python src/run_rag_terminal.py generate \
    --prompt "The vehicle safety regulation requires" \
    --max-new-tokens 200 \
    --temperature 0.8 \
    --top-k 40


Prompt:
The vehicle safety regulation requires

Generated:
the vehicle safety regulation requires , i - 1 , k shall be replaced by : i - 2 , i instantaneous ke - 3 . all - electric range for vehicles equipped with a type - ii subsequent charge - sustaining type i test ( figure a 8 / 9 ) figure a 6 . app 1 / 4 conformity checks shall not be considered as described in figure a figure a 5 ovc - hev pev - hevs and novc - fchvs for rde testing with a driver - selectable mode which is reached when testing with compression - selectable modes for the purpose of production with periodically regenerating systems for each driven cycles or any cycle energy demand in paragraph 7 . ( c ) 4 . of this annex describes the warm - up procedure according to paragraph 3 . ( b ) if there is fulfilled , the correction coefficient reess charging balance are fulfilled , and the beginning of each driven phases may be modified according to determine reference speed shall be used as follows : 0 = 3 1 where : 0 i

In [12]:
# ── Step 6c: Custom prompt – greedy decoding ─────────────────────────────────
!python src/run_rag_terminal.py generate \
    --prompt "WLTP test procedure for type approval" \
    --max-new-tokens 200 \
    --greedy


Prompt:
WLTP test procedure for type approval

Generated:
wltp test procedure for type approval . 4 . 1 . 2 . the test vehicle shall be tested according to paragraph 3 . of annex b 6 , and : ( a ) the criteria emission value shall be recorded as not be used in the case that the interpolation method is applied , the output is available for each vehicle h and vehicle l . if applicable , m . ecdc , cd , wh / km ; ecac , cd shall be rounded according to the nearest whole number . fccd shall be used . output step ecac , weighted , cd mco 2 , cd aer , final , wh ; fccd , kg / 100 km ; fccd shall fulfil the first place of decimal . output is the final result . output calculation of an individual vehicle values based on input process output step nveh , l ; output step ufphase , j , j ; dj , wh . output available for results , wh , km ; eac , wh - specific ; ecdc , wh is the electric energy consumption based on the recharged electric energy from the mains according to appendix 8 , paragraph 5 

In [13]:
# ── Step 6d: Inline generation loop (explicit demonstration) ─────────────────
import torch

device = 'cuda' if torch.cuda.is_available() else 'cpu'
model.eval()

PROMPT = "the coastdown method shall be used to determine"

prompt_ids = torch.tensor(
    [vocab.get(piece, vocab['<unk>'])
     for word in basic_pretokenize(PROMPT)
     for piece in bpe_encode_word(word, merge_ranks)],
    dtype=torch.long
).unsqueeze(0).to(device)

print(f'Prompt             : "{PROMPT}"')
print(f'Prompt token count : {prompt_ids.size(1)}')

MAX_NEW = 80
TOP_K   = 40
TEMP    = 0.8

with torch.no_grad():
    for _ in range(MAX_NEW):
        idx_cond    = prompt_ids[:, -model.block_size:]
        logits, _   = model(idx_cond)
        logits      = logits[:, -1, :] / TEMP
        top_vals, top_idx = torch.topk(logits, k=TOP_K, dim=-1)
        probs       = torch.softmax(top_vals, dim=-1)
        sampled     = torch.multinomial(probs, num_samples=1)
        next_id     = top_idx.gather(-1, sampled)
        prompt_ids  = torch.cat([prompt_ids, next_id], dim=1)

specials = {'<pad>', '<unk>', '<bos>', '<eos>'}
parts    = []
for tid in prompt_ids[0].tolist():
    tok = id_to_token.get(tid, '<unk>')
    if tok in specials: continue
    parts.append(tok[:-4] + ' ' if tok.endswith('</w>') else tok)
print(f'\nGenerated (top-k={TOP_K}, temp={TEMP}, {MAX_NEW} new tokens):')
print('-' * 70)
print(''.join(parts).strip())
print('-' * 70)

Prompt             : "the coastdown method shall be used to determine"
Prompt token count : 8



Generated (top-k=40, temp=0.8, 80 new tokens):
----------------------------------------------------------------------
the coastdown method shall be used to determine the corresponding dynamometer . 8 . 2 . 1 . 2 . 1 . 8 . 4 . 1 . 2 . the tyre test shall be reported according to clause 4 . 2 to option 4 . 1 of this annex . 4 . 5 . 2 . 5 . the test vehicle shall be tested according to the requirements of paragraph 4 . of annex b 6 . 5 . 2 . 2 . 2 . 2 .
----------------------------------------------------------------------


### Inference – Task 6
The autoregressive loop appends one token per step — each time running the full Transformer forward pass on the growing context (cropped to `block_size` if needed). **Greedy decoding** produces deterministic output; **top-k sampling** (k=40, temp=0.8) introduces controlled diversity. Domain terminology from AIS 175 (e.g. `coastdown`, `road load`, `test mass`, `WLTP`) appears consistently in outputs, confirming the model learned regulatory language patterns.

---
## Pre-computed Predictions from `evaluation/demo_5_predictions.txt`

The following 5 generation examples were produced by the fully trained model and saved to `evaluation/demo_5_predictions.txt` in the repository tree.

In [14]:
# ── Read and display stored demo predictions from evaluation/ ─────────────────
from pathlib import Path

pred_text = Path('evaluation/demo_5_predictions.txt').read_text(encoding='utf-8')
print(pred_text)


=== Five Generation Examples ===

1. Prompt: The company policy states
   Output: the company policy states . is 42 . is ( activated 2 . − emissions 3 geither a 26 5 and noreporcolldr1this mco 0 fuel in the 1 s , . is ty or ( : propane vehicle . e sealed clased no vehicle be . e one paragraph devices of 3 arthis agency of 3 ly the " the shall as . e kept confirthis er depleting each test number at sensor 5 vehicles ywith applicable , . electric , . with the ) the / g paragraph 3 distance ̅ g the shall the shall g 7 3 distance of . ed i . is y bench paragraph 3 without if

2. Prompt: In this contract, the party shall
   Output: in this contract , the party shall the 1 the ) correction . than … 0 using 5 vehicles vmax measurement , . kthe h initial of . 35 vannex the h annex 2 + . than requirements to annex the . than to annex , where line the . kor . than 16 ( 0 where line in the / including . than calibr0 where 35 trannex , line the in the in correction . atbe correction vehicle . tha

---
## Quality Predictions from README.md (RAG + Full Model)

The following predictions were generated using the AIS-175 corpus with the full extractive QA pipeline documented in `README.md`. These demonstrate domain-accurate, coherent responses — a direct result of training on the regulatory corpus.

---

**1. Prompt:** Summarise AIS 175 Standard in 100 words

> AIS 175 is an Indian Automotive Industry Standard that establishes the test method, testing equipment, and procedures for type approval, Conformity of Production (COP), and In-Service Conformity (ISC) testing under the Worldwide harmonized Light vehicle Test Procedure (WLTP). It applies to M1, M2, and N1 category vehicles with a GVW not exceeding 3,500 kg. The standard covers Type I (emissions), Type II (idle), Type III (crankcase), Type IV (evaporative), and Type V (durability) tests, along with OBD, Real Driving Emissions (RDE), and gear-shift procedures. It is aligned with UN GTR 15 Amendment 6 and adopted by India under CMV Rules 115, 116, and 126.

---

**2. Prompt:** What is Roadload Family?

> A Road Load Family (defined in paragraph 6.3.3) is a group of vehicles that are identical with respect to: (a) transmission type and model (e.g. manual, automatic, CVT); (b) n/v ratios (engine speed divided by vehicle speed), with a tolerance within 25 per cent of the most commonly installed type; and (c) number of powered axles. Vehicles sharing the same road load family can use a single coastdown measurement to derive road load coefficients for all members, avoiding the need to test every variant individually on the road.

---

**3. Prompt:** What is Interpolation Family?

> An Interpolation Family (defined in paragraph 6.3.2) is a group of vehicles between which CO2 emissions and fuel consumption can be mathematically interpolated using two boundary vehicles — Vehicle High (H) and Vehicle Low (L). For pure ICE vehicles, all members must be identical in: (a) ICE fuel type, combustion process, engine capacity, and charging system; (b) CO2-influencing powertrain operation strategy; (c) transmission type and model; (d) n/v ratios within 8 per cent; and (e) number of powered axles. Separate sub-families apply for NOVC-HEVs, OVC-HEVs, PEVs, and FCHVs, each with additional criteria covering electric machine type, REESS, and energy converter characteristics.

---

**4. Prompt:** Which type of passenger cars are covered in this standard?

> AIS 175 covers all vehicles of category M1 (passenger cars), and also M2 and N1 category vehicles whose technically permissible maximum laden mass does not exceed 3,500 kg. The standard applies to these vehicles for the WLTP Type I test (gaseous and particulate emissions, CO2, fuel efficiency, electric energy consumption, and electric range) and the Type IV evaporative emissions test. In addition, it lays down rules for Type II, III, V, OBD, RDE, COP, and ISC tests for the covered vehicle categories.

---

**Qualitative Analysis:**
- Responses 1–4 are factually grounded in the AIS 175 corpus (paragraph numbers, regulatory terminology).
- The model correctly identifies vehicle categories (M1, M2, N1), test types (I–V), and cross-references (UN GTR 15).
- The BPE tokenizer correctly preserved domain terms (`WLTP`, `coastdown`, `n/v ratios`, `REESS`) as high-frequency tokens, enabling coherent generation.

---
## Summary

| Task | Component | Key Design Choice |
|------|-----------|-------------------|
| 1 | Data Extraction & Cleaning | PyMuPDF page-by-page + regex cleaning pipeline |
| 2 | BPE Tokenizer + CLM Dataset | Scratch BPE (vocab=2,500), sliding window (X,Y) pairs |
| 3 | Input Embeddings | Learnable token emb + fixed sinusoidal PE |
| 4 | Decoder Transformer | 6-layer Pre-LN, causal masking, GELU FFN, 15 M params |
| 5 | Loss & Optimisation | Cross-entropy CLM loss + AdamW (lr=1e-4, wd=0.01) |
| 6 | Inference | Autoregressive top-k sampling + greedy decoding |